Registry/RegistryLog.py

```diff
@@ -0,0 +1,150 @@
+        """
+        try:
+            self._log_buf = filelikeobject_log.read()
+        except AttributeError:
+                with open(filelikeobject_log, "rb") as f:
+                    self._log_buf = f.read()
+
+        self._regf = RegistryParse.REGFBlock(self._log_buf, 0, False)
+
+        if self._regf.is_old_transaction_log_file():
+            raise RegistryParse.NotSupportedException("Old transaction log files are not supported")
+
+        if self._regf.is_primary_file():
+            raise RegistryParse.ParseException("Cannot load a primary file as a transaction log file")
+
+        if not self._regf.is_new_transaction_log_file():
+            raise RegistryParse.NotSupportedException("Unknown file type")
+
+        if self._regf.clustering_factor() != 1:
+            raise RegistryParse.NotSupportedException("Clustering factor not equal to 1 is not supported")
+
+        recover_header, recover_data = self._regf.recovery_required()
+        if recover_header or recover_data:
+            raise RegistryParse.NotSupportedException("This transaction log file requires self-healing")
+
+        self._primary_buf = filelikeobject_primary.read(512)
+        self._primary = filelikeobject_primary
+
+        self._primary_regf = RegistryParse.REGFBlock(self._primary_buf, 0, False)
+        self._hive_flags = None
+        self._hive_sequence = None
+
+    def latest_hive_flags(self):
+        """Return the latest hive flags. At present, only one bit mask (0x1) is used."""
+        return self._hive_flags
+
+    def latest_hive_sequence(self):
+        """Return the latest hive_sequence1 (the same as hive_sequence2 after recovery)."""
+        return self._hive_sequence
+
+    def first_log_sequence(self):
+        """Returns the first log sequence number."""
+        return self._regf.hive_sequence2()
+
+    def is_eligible_log(self):
+        """Check if this log is eligible for the primary file."""
+        if not self._primary_regf.validate_checksum():
+            return True
+
+        if self.first_log_sequence() >= self._primary_regf.hive_sequence2():
+            return True
+
+        return False
+
+    def is_starting_log(self, another_log_file):
+        """
+        When the dual-logging scheme is used, check if this log shall be applied first.
+        Another RegistryLog instance is checked against this one.
+        """
+        another_seqnum = another_log_file.first_log_sequence()
+        this_seqnum = self.first_log_sequence()
+        if this_seqnum >= another_seqnum:
+            delta = this_seqnum - another_seqnum
+            starting = False
+        else:
+            delta = another_seqnum - this_seqnum
+            starting = True
+
+        if c_uint32(delta).value <= 0x7FFFFFFF:
+            return starting
+        else:
+            # Sequence numbers did overflow.
+            return not starting
+
+    def write_dirty_page(self, dirty_page_reference, dirty_page):
```

Comment:
it looks to me like this writes to the original file. this is important to document for analysts using a script on extracted evidence. it might be a surprise if the underlying files changed without warning. if nothing else, the scripts that recover a hive from the transaction log should log a status message to this effect.

alternatively, sample scripts could use `StringIO` instances as the `primary` file to ensure that the source file remains read-only.